src/pass/lift_if_then_else.cc

```diff
@@ -0,0 +1,403 @@
+ *            if (likely(j > 2))
+ *                A[i+j+k] = B[i+j+k]
+ *
+ */
+class IfThenElseLifter {
+ public:
+  Stmt VisitAndMutate(const Stmt& stmt) {
+    GenerateInternalData(stmt);
+    return PostOrderMutate(stmt);
+  }
+
+ private:
+  void GenerateInternalData(const Stmt& stmt);
+  Stmt PostOrderMutate(const Stmt& stmt);
+  size_t GetUpdatedFor(const Stmt& for_stmt, const Stmt& if_stmt);
+  Stmt LiftIf(const Stmt& if_stmt);
+
+  LifterMap if2for_map_;
+  LifterMap top_for_var_map_;
+  LifterMap for_tracking_map_;
+  LifterMap for2if_map_;
+  VarMap cond_var_map_;
+  std::vector<Stmt> ordered_for_list_;
+};
+
+// Check whether a given IfThenElse stmt is the first one appearing
+// in a For stmt.
+bool is_first_if(const Stmt& for_stmt, const Stmt& if_stmt) {
+  std::vector<size_t> if_hash_list;
+
+  PostOrderVisit(for_stmt.as<For>()->body, [&](const NodeRef& node) {
+    if (node.as<IfThenElse>()) {
+      if_hash_list.push_back(node.hash());
+    }
+  });
+  return if_hash_list.empty() ? false : if_stmt.hash() == if_hash_list.back();
+}
+
+// Update upper level For node when current For node is modified.
+// With this function we only need to visit and mutate top level For node
+// in the main VisitAndMutate function.
+Stmt update_for(const Stmt& parent_for_stmt, const Stmt& new_if_stmt) {
+  std::vector<size_t> for_hash_list;
+
+  PostOrderVisit(parent_for_stmt.as<For>()->body, [&](const NodeRef& node) {
+    if (node.as<For>()) {
+      for_hash_list.push_back(node.hash());
+    }
+  });
+
+  PackedFunc replace_target_for = PackedFunc(
+    [&](TVMArgs args, TVMRetValue *ret){
+      const NodeRef& current_for = args[0];
+      if (current_for.hash() == for_hash_list.back()) {
+        *ret = new_if_stmt;
+      }
+    });
+
+  return IRTransform(parent_for_stmt, nullptr, replace_target_for,
+                     {Expr("For")});
+}
+
+// Remove IfThenElse node from a For node.
+// A pair of For nodes will be generated.
+std::pair<Stmt, Stmt> RemoveIf(const Stmt& for_stmt, const Stmt& if_stmt) {
+  Stmt then_for;
+  Stmt else_for;
+
+  PackedFunc replace_then_case = PackedFunc(
+    [&](TVMArgs args, TVMRetValue *ret){
+      const NodeRef& node  = args[0];
+      if (node == if_stmt) {
+        *ret = node.as<IfThenElse>()->then_case;
+      }
+    });
+
+  PackedFunc replace_else_case = PackedFunc(
+    [&](TVMArgs args, TVMRetValue *ret){
+      const NodeRef& node  = args[0];
+      if (node == if_stmt) {
+        *ret = node.as<IfThenElse>()->else_case;
+      }
+    });
+
+  then_for = IRTransform(for_stmt, nullptr, replace_then_case,
+                         {Expr("IfThenElse")});
+  if (if_stmt.as<IfThenElse>()->else_case) {
+    else_for = IRTransform(for_stmt, nullptr, replace_else_case,
+                           {Expr("IfThenElse")});
+  }
+
+  return std::make_pair(then_for, else_for);
+}
+
+// Generate internal data structures for lifter.
+void IfThenElseLifter::GenerateInternalData(const Stmt& stmt) {
```

Comment:
can we separate the function to smaller units and make the function name more meaningful?